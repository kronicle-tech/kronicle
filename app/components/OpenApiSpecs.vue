<template>
  <ul v-if="component.openApiSpecs && component.openApiSpecs.length > 0">
    <li
      v-for="(openApiSpec, openApiSpecIndex) in component.openApiSpecs"
      :key="openApiSpecIndex"
    >
      <b-link
        v-if="openApiSpec.spec"
        :href="`/components/${componentId}/openapi-specs/${
          openApiSpecIndex + 1
        }`"
      >
        {{ openApiSpec.url || openApiSpec.file }}
      </b-link>
      <div v-else>
        {{ openApiSpec.url || openApiSpec.file }}
        <b-badge variant="danger">Missing</b-badge>
      </div>
      <Markdown :markdown="openApiSpec.description" />
    </li>
  </ul>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'
import { BBadge, BLink } from 'bootstrap-vue'
import {
  Component,
  OpenApiSpec,
  OpenApiSpecsState,
} from '~/types/kronicle-service'
import Markdown from '~/components/Markdown.vue'
import { findComponentState } from '~/src/componentStateUtils'

export default Vue.extend({
  components: {
    'b-link': BLink,
    'b-badge': BBadge,
    Markdown,
  },
  props: {
    component: {
      type: Object as PropType<Component>,
      default: undefined,
    },
  },
  computed: {
    openApiSpecs(): OpenApiSpec[] {
      const openApiSpecs: OpenApiSpecsState | undefined = findComponentState(
        this.component,
        'openapi-specs'
      )
      return openApiSpecs?.openApiSpecs ?? []
    },
  },
})
</script>
