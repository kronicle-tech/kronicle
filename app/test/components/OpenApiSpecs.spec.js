import { mount } from '@vue/test-utils'
import OpenApiSpecs from '@/components/OpenApiSpecs.vue'

describe('OpenApiSpecs', () => {
  let propsData
  let wrapper
  const createWrapper = () => {
    wrapper = mount(OpenApiSpecs, { propsData })
  }

  beforeEach(() => {
    propsData = {}
  })

  afterEach(() => {
    wrapper.destroy()
    wrapper = null
  })

  describe('when openApiSpecs array is not set', () => {
    beforeEach(() => {
      propsData.component = {
        id: 'test-component-id',
      }
    })

    test('renders nothing', () => {
      createWrapper()
      expect(wrapper.html()).toEqual('')
    })
  })

  describe('when openApiSpecs array is set to an empty array', () => {
    beforeEach(() => {
      propsData.component = {
        id: 'test-component-id',
        openApiSpecs: [],
      }
    })

    test('renders nothing', () => {
      createWrapper()
      expect(wrapper.html()).toEqual('')
    })
  })

  describe('when openApiSpecs array is set to one OpenAPI spec', () => {
    beforeEach(() => {
      propsData.component = {
        id: 'test-component-id',
        openApiSpecs: [
          {
            scannerId: 'test-scanner-1',
            url: 'https://example.com/',
            description: 'Test Description 1',
            spec: {},
          },
        ],
      }
    })

    test('renders an unordered list showing the OpenAPI spec', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when openApiSpecs array is set to an OpenAPI spec with no spec', () => {
    beforeEach(() => {
      propsData.component = {
        id: 'test-component-id',
        openApiSpecs: [
          {
            scannerId: 'test-scanner-1',
            url: 'https://example.com/',
            description: 'Test Description 1',
          },
        ],
      }
    })

    test('renders the OpenAPI spec with a `missing` badge', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when openApiSpecs array is set to an OpenAPI spec with a url', () => {
    beforeEach(() => {
      propsData.component = {
        id: 'test-component-id',
        openApiSpecs: [
          {
            scannerId: 'test-scanner-1',
            url: 'https://example.com/',
            description: 'Test Description 1',
            spec: {},
          },
        ],
      }
    })

    test('renders the OpenAPI spec showing the URL as the spec name', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when openApiSpecs array is set to an OpenAPI spec with a file', () => {
    beforeEach(() => {
      propsData.component = {
        id: 'test-component-id',
        openApiSpecs: [
          {
            scannerId: 'test-scanner-1',
            file: 'test/file.yaml',
            description: 'Test Description 1',
            spec: {},
          },
        ],
      }
    })

    test('renders the OpenAPI spec showing the file as the spec name', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when openApiSpecs array is set to an OpenAPI spec with a description', () => {
    beforeEach(() => {
      propsData.component = {
        id: 'test-component-id',
        openApiSpecs: [
          {
            scannerId: 'test-scanner-1',
            url: 'https://example.com/',
            description: 'Test Description 1',
            spec: {},
          },
        ],
      }
      propsData.componentId = 'test-component-id'
    })

    test('renders the OpenAPI spec with a description', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when openApiSpecs array is set to an OpenAPI spec with no description', () => {
    beforeEach(() => {
      propsData.component = {
        id: 'test-component-id',
        openApiSpecs: [
          {
            scannerId: 'test-scanner-1',
            url: 'https://example.com/',
            spec: {},
          },
        ],
      }
    })

    test('renders the OpenAPI spec without a description', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when openApiSpecs array is set to one OpenAPI spec with Markdown in the description', () => {
    beforeEach(() => {
      propsData.component = {
        id: 'test-component-id',
        openApiSpecs: [
          {
            scannerId: 'test-scanner-1',
            url: 'https://example.com/',
            description: 'Text with *bold* formatting',
            spec: {},
          },
        ],
      }
    })

    test('renders the OpenAPI spec with the Markdown converted to HTML', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when openApiSpecs array is set to multiple OpenAPI specs', () => {
    beforeEach(() => {
      propsData.component = {
        id: 'test-component-id',
        openApiSpecs: [
          {
            scannerId: 'test-scanner-2',
            url: 'https://example.com/',
            description: 'Test Description 2',
            spec: {},
          },
          {
            scannerId: 'test-scanner-2',
            url: 'https://example.com/',
            description: 'Test Description 2',
            spec: {},
          },
        ],
      }
    })

    test('renders an unordered list showing the OpenAPI specs', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })
})
