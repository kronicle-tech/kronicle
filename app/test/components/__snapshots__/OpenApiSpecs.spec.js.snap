// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenApiSpecs when openApiSpecs array is set to an OpenAPI spec with a description renders the OpenAPI spec with a description 1`] = `
"<ul componentid=\\"test-component-id\\">
  <li><a href=\\"/components/undefined/openapi-specs/1\\" target=\\"_self\\" class=\\"\\">
      https://example.com/
    </a>
    <div>
      <p>Test Description 1</p>
    </div>
  </li>
</ul>"
`;

exports[`OpenApiSpecs when openApiSpecs array is set to an OpenAPI spec with a file renders the OpenAPI spec showing the file as the spec name 1`] = `
"<ul>
  <li><a href=\\"/components/undefined/openapi-specs/1\\" target=\\"_self\\" class=\\"\\">
      test/file.yaml
    </a>
    <div>
      <p>Test Description 1</p>
    </div>
  </li>
</ul>"
`;

exports[`OpenApiSpecs when openApiSpecs array is set to an OpenAPI spec with a url renders the OpenAPI spec showing the URL as the spec name 1`] = `
"<ul>
  <li><a href=\\"/components/undefined/openapi-specs/1\\" target=\\"_self\\" class=\\"\\">
      https://example.com/
    </a>
    <div>
      <p>Test Description 1</p>
    </div>
  </li>
</ul>"
`;

exports[`OpenApiSpecs when openApiSpecs array is set to an OpenAPI spec with no description renders the OpenAPI spec without a description 1`] = `
"<ul>
  <li><a href=\\"/components/undefined/openapi-specs/1\\" target=\\"_self\\" class=\\"\\">
      https://example.com/
    </a>
    <div></div>
  </li>
</ul>"
`;

exports[`OpenApiSpecs when openApiSpecs array is set to an OpenAPI spec with no spec renders the OpenAPI spec with a \`missing\` badge 1`] = `
"<ul>
  <li>
    <div>
      https://example.com/
      <span class=\\"badge badge-danger\\">Missing</span></div>
    <div>
      <p>Test Description 1</p>
    </div>
  </li>
</ul>"
`;

exports[`OpenApiSpecs when openApiSpecs array is set to multiple OpenAPI specs renders an unordered list showing the OpenAPI specs 1`] = `
"<ul>
  <li><a href=\\"/components/undefined/openapi-specs/1\\" target=\\"_self\\" class=\\"\\">
      https://example.com/
    </a>
    <div>
      <p>Test Description 2</p>
    </div>
  </li>
  <li><a href=\\"/components/undefined/openapi-specs/2\\" target=\\"_self\\" class=\\"\\">
      https://example.com/
    </a>
    <div>
      <p>Test Description 2</p>
    </div>
  </li>
</ul>"
`;

exports[`OpenApiSpecs when openApiSpecs array is set to one OpenAPI spec renders an unordered list showing the OpenAPI spec 1`] = `
"<ul>
  <li><a href=\\"/components/undefined/openapi-specs/1\\" target=\\"_self\\" class=\\"\\">
      https://example.com/
    </a>
    <div>
      <p>Test Description 1</p>
    </div>
  </li>
</ul>"
`;

exports[`OpenApiSpecs when openApiSpecs array is set to one OpenAPI spec with Markdown in the description renders the OpenAPI spec with the Markdown converted to HTML 1`] = `
"<ul>
  <li><a href=\\"/components/undefined/openapi-specs/1\\" target=\\"_self\\" class=\\"\\">
      https://example.com/
    </a>
    <div>
      <p>Text with <em>bold</em> formatting</p>
    </div>
  </li>
</ul>"
`;
